File: qqbot-module-qqbot-runtime/src/main/java/com/mieai/qqbot/admin/bot/BotMessageAdministrationService.java
package com.mieai.qqbot.admin.bot;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.mieai.qqbot.client.MediaAssetStore;
import com.mieai.qqbot.client.MediaAsset;
import com.mieai.qqbot.client.QqMediaMessageRequest;
import com.mieai.qqbot.client.QqMessageTargetType;
import com.mieai.qqbot.client.QqRichMessageKind;
import com.mieai.qqbot.client.QqRichMessageRequest;
import com.mieai.qqbot.client.QqTextMessageRequest;
import com.mieai.qqbot.domain.bot.BotId;
import com.mieai.qqbot.persistence.bot.BotRepository;
import com.mieai.qqbot.persistence.bot.StoredBot;
import com.mieai.qqbot.persistence.outbox.NewOutboxJob;
import com.mieai.qqbot.persistence.outbox.OutboxRepository;
import com.mieai.qqbot.runtime.outbox.OutboundMediaPayload;
import com.mieai.qqbot.runtime.outbox.OutboundRichPayload;
import com.mieai.qqbot.runtime.outbox.OutboundTextPayload;
import java.net.URI;
import java.time.Clock;
import java.time.Instant;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.UUID;
import org.springframework.http.HttpStatus;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class BotMessageAdministrationService {
    private final BotRepository bots;
    private final OutboxRepository outbox;
    private final MediaAssetStore mediaStore;
    private final ObjectMapper mapper;
    private final Clock clock;

    @Autowired
    public BotMessageAdministrationService(BotRepository bots, OutboxRepository outbox,
            MediaAssetStore mediaStore, ObjectMapper mapper) {
        this(bots, outbox, mediaStore, mapper, Clock.systemUTC());
    }

    BotMessageAdministrationService(BotRepository bots, OutboxRepository outbox,
            MediaAssetStore mediaStore, ObjectMapper mapper, Clock clock) {
        this.bots = Objects.requireNonNull(bots, "bots must not be null");
        this.outbox = Objects.requireNonNull(outbox, "outbox must not be null");
        this.mediaStore = Objects.requireNonNull(mediaStore, "mediaStore must not be null");
        this.mapper = Objects.requireNonNull(mapper, "mapper must not be null");
        this.clock = Objects.requireNonNull(clock, "clock must not be null");
    }

    public BotMessageResponse send(String botId, SendBotMessageRequest request) {
        BotId id = BotId.parse(botId);
        StoredBot bot = bots.findById(id).orElseThrow(() -> new BotMessageAdministrationException(
                HttpStatus.NOT_FOUND, "BOT_NOT_FOUND", "Bot does not exist"));
        if (!bot.definition().enabled()) throw new BotMessageAdministrationException(
                HttpStatus.CONFLICT, "BOT_DISABLED", "Bot is disabled");
        Objects.requireNonNull(request, "request must not be null");
        Instant now = clock.instant();
        UUID jobId = UUID.randomUUID();
        String type;
        String payload;
        MediaAsset stagedAsset = null;
        try {
            switch (request.kind()) {
                case TEXT -> {
                    requireText(request.content());
                    new QqTextMessageRequest(request.targetType(), request.targetId(), request.content(),
                            Optional.ofNullable(request.replyMessageId()), Optional.ofNullable(request.replyEventId()),
                            request.messageSequence());
                    type = OutboundTextPayload.JOB_TYPE;
                    payload = mapper.writeValueAsString(new OutboundTextPayload(
                            request.targetType(), request.targetId(), request.content(),
                            request.replyMessageId(), request.replyEventId(), request.messageSequence()));
                }
                case MEDIA -> {
                    if (request.mediaKind() == null) throw invalid("mediaKind is required");
                    if ((request.targetType() == QqMessageTargetType.CHANNEL
                            || request.targetType() == QqMessageTargetType.DIRECT)
                            && request.mediaKind() != com.mieai.qqbot.client.QqMediaKind.IMAGE) {
                        throw invalid("Channel and direct messages only support image media");
                    }
                    UUID assetId = request.mediaAssetId();
                    String url = request.mediaUrl();
                    if (assetId == null && (url == null || url.isBlank())) throw invalid("mediaUrl or mediaAssetId is required");
                    if (assetId != null) {
                        MediaAsset asset = mediaStore.find(id, assetId)
                                .orElseThrow(() -> invalid("mediaAssetId is not available"));
                        stagedAsset = asset;
                        if (asset.sizeBytes() > bot.definition().maxMediaUploadBytes()) {
                            throw new MediaUploadTooLargeException(asset.sizeBytes(),
                                    bot.definition().maxMediaUploadBytes());
                        }
                        if (asset.kind() != request.mediaKind()) {
                            throw invalid("mediaAssetId does not match mediaKind");
                        }
                        url = "https://asset.invalid/" + assetId;
                    }
                    new QqMediaMessageRequest(request.targetType(), request.targetId(), request.mediaKind(),
                            URI.create(url), Optional.ofNullable(request.content()),
                            Optional.ofNullable(request.replyMessageId()), Optional.ofNullable(request.replyEventId()),
                            request.messageSequence());
                    type = OutboundMediaPayload.JOB_TYPE;
                    payload = mapper.writeValueAsString(new OutboundMediaPayload(
                            request.targetType(), request.targetId(), request.mediaKind(), url,
                            request.content(), request.replyMessageId(), request.replyEventId(),
                            request.messageSequence(), assetId));
                }
                default -> {
                    Map<String, Object> body = request.payload() == null ? Map.of() : request.payload();
                    if (body.isEmpty()) throw invalid("payload is required for rich messages");
                    new QqRichMessageRequest(request.targetType(), request.targetId(),
                            QqRichMessageKind.valueOf(request.kind().name()), body,
                            Optional.ofNullable(request.replyMessageId()), Optional.ofNullable(request.replyEventId()),
                            request.messageSequence());
                    type = OutboundRichPayload.JOB_TYPE;
                    payload = mapper.writeValueAsString(new OutboundRichPayload(
                            request.targetType(), request.targetId(),
                            QqRichMessageKind.valueOf(request.kind().name()), body,
                            request.replyMessageId(), request.replyEventId(), request.messageSequence()));
                }
            }
        } catch (JsonProcessingException exception) {
            deleteQuietly(stagedAsset);
            throw new BotMessageAdministrationException(HttpStatus.BAD_REQUEST, "INVALID_MESSAGE", "Message payload is invalid");
        } catch (IllegalArgumentException | NullPointerException exception) {
            deleteQuietly(stagedAsset);
            throw invalid(exception.getMessage() == null ? "Message payload is invalid" : exception.getMessage());
        } catch (RuntimeException exception) {
            deleteQuietly(stagedAsset);
            throw exception;
        }
        try {
            outbox.create(new NewOutboxJob(jobId, bot.definition().environment(), id, Optional.empty(),
                    type, Optional.empty(), payload, now, now));
        } catch (RuntimeException exception) {
            deleteQuietly(stagedAsset);
            throw exception;
        }
        return new BotMessageResponse(jobId, false, now);
    }

    private static void requireText(String value) {
        if (value == null || value.isBlank()) throw invalid("content is required");
    }

    private static BotMessageAdministrationException invalid(String message) {
        return new BotMessageAdministrationException(HttpStatus.BAD_REQUEST, "INVALID_MESSAGE", message);
    }

    private void deleteQuietly(MediaAsset asset) {
        if (asset == null) return;
        try {
            mediaStore.delete(asset);
        } catch (RuntimeException ignored) {
            // Preserve the request failure; the staged directory remains operator-recoverable.
        }
    }
}

File: qqbot-module-qqbot-runtime/src/main/kotlin/com/mieai/qqbot/admin/bot/BotMessageAdministrationService.kt
package com.mieai.qqbot.admin.bot

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.mieai.qqbot.client.MediaAsset
import com.mieai.qqbot.client.MediaAssetStore
import com.mieai.qqbot.client.QqMediaKind
import com.mieai.qqbot.client.QqMediaMessageRequest
import com.mieai.qqbot.client.QqMessageTargetType
import com.mieai.qqbot.client.QqRichMessageKind
import com.mieai.qqbot.client.QqRichMessageRequest
import com.mieai.qqbot.client.QqTextMessageRequest
import com.mieai.qqbot.domain.bot.BotId
import com.mieai.qqbot.persistence.bot.BotRepository
import com.mieai.qqbot.persistence.outbox.NewOutboxJob
import com.mieai.qqbot.persistence.outbox.OutboxRepository
import com.mieai.qqbot.runtime.outbox.OutboundMediaPayload
import com.mieai.qqbot.runtime.outbox.OutboundRichPayload
import com.mieai.qqbot.runtime.outbox.OutboundTextPayload
import java.net.URI
import java.time.Clock
import java.util.UUID
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service

@Service
class BotMessageAdministrationService(
    private val bots: BotRepository,
    private val outbox: OutboxRepository,
    private val mediaStore: MediaAssetStore,
    private val mapper: ObjectMapper,
    private val clock: Clock = Clock.systemUTC(),
) {
    fun send(botId: String, request: SendBotMessageRequest): BotMessageResponse {
        val id = BotId.parse(botId)
        val bot = bots.findById(id) ?: throw BotMessageAdministrationException(
            HttpStatus.NOT_FOUND,
            "BOT_NOT_FOUND",
            "Bot does not exist",
        )
        if (!bot.definition.enabled) {
            throw BotMessageAdministrationException(HttpStatus.CONFLICT, "BOT_DISABLED", "Bot is disabled")
        }
        val now = clock.instant()
        val jobId = UUID.randomUUID()
        var stagedAsset: MediaAsset? = null
        val encoded = try {
            val targetType = requireNotNull(request.targetType) { "targetType must not be null" }
            val targetId = requireNotNull(request.targetId) { "targetId must not be null" }
            val sequence = request.messageSequence ?: 1
            val messageKind = request.kind
            when (messageKind) {
                BotMessageKind.TEXT -> {
                    val content = request.content
                    requireText(content)
                    QqTextMessageRequest(
                        targetType,
                        targetId,
                        content!!,
                        request.replyMessageId,
                        request.replyEventId,
                        sequence,
                    )
                    EncodedMessage(
                        OutboundTextPayload.JOB_TYPE,
                        mapper.writeValueAsString(
                            OutboundTextPayload(
                                targetType,
                                targetId,
                                content,
                                request.replyMessageId,
                                request.replyEventId,
                                sequence,
                            ),
                        ),
                    )
                }
                BotMessageKind.MEDIA -> {
                    val mediaKind = request.mediaKind ?: throw invalid("mediaKind is required")
                    if ((targetType == QqMessageTargetType.CHANNEL || targetType == QqMessageTargetType.DIRECT) &&
                        mediaKind != QqMediaKind.IMAGE
                    ) {
                        throw invalid("Channel and direct messages only support image media")
                    }
                    val assetId = request.mediaAssetId
                    var url = request.mediaUrl
                    if (assetId == null && url.isNullOrBlank()) {
                        throw invalid("mediaUrl or mediaAssetId is required")
                    }
                    if (assetId != null) {
                        val asset = mediaStore.find(id, assetId)
                            ?: throw invalid("mediaAssetId is not available")
                        stagedAsset = asset
                        if (asset.sizeBytes > bot.definition.maxMediaUploadBytes) {
                            throw MediaUploadTooLargeException(
                                asset.sizeBytes,
                                bot.definition.maxMediaUploadBytes,
                            )
                        }
                        if (asset.kind != mediaKind) throw invalid("mediaAssetId does not match mediaKind")
                        url = "https://asset.invalid/$assetId"
                    }
                    val resolvedUrl = requireNotNull(url)
                    QqMediaMessageRequest(
                        targetType,
                        targetId,
                        mediaKind,
                        URI.create(resolvedUrl),
                        request.content,
                        request.replyMessageId,
                        request.replyEventId,
                        sequence,
                    )
                    EncodedMessage(
                        OutboundMediaPayload.JOB_TYPE,
                        mapper.writeValueAsString(
                            OutboundMediaPayload(
                                targetType,
                                targetId,
                                mediaKind,
                                resolvedUrl,
                                request.content,
                                request.replyMessageId,
                                request.replyEventId,
                                sequence,
                                assetId,
                            ),
                        ),
                    )
                }
                BotMessageKind.MARKDOWN,
                BotMessageKind.KEYBOARD,
                BotMessageKind.ARK,
                BotMessageKind.EMBED,
                -> {
                    val body = request.payload ?: emptyMap()
                    if (body.isEmpty()) throw invalid("payload is required for rich messages")
                    val kind = QqRichMessageKind.valueOf(messageKind.name)
                    QqRichMessageRequest(
                        targetType,
                        targetId,
                        kind,
                        body,
                        request.replyMessageId,
                        request.replyEventId,
                        sequence,
                    )
                    EncodedMessage(
                        OutboundRichPayload.JOB_TYPE,
                        mapper.writeValueAsString(
                            OutboundRichPayload(
                                targetType,
                                targetId,
                                kind,
                                body,
                                request.replyMessageId,
                                request.replyEventId,
                                sequence,
                            ),
                        ),
                    )
                }
                null -> throw NullPointerException("kind must not be null")
            }
        } catch (_: JsonProcessingException) {
            deleteQuietly(stagedAsset)
            throw BotMessageAdministrationException(
                HttpStatus.BAD_REQUEST,
                "INVALID_MESSAGE",
                "Message payload is invalid",
            )
        } catch (exception: IllegalArgumentException) {
            deleteQuietly(stagedAsset)
            throw invalid(exception.message ?: "Message payload is invalid")
        } catch (exception: NullPointerException) {
            deleteQuietly(stagedAsset)
            throw invalid(exception.message ?: "Message payload is invalid")
        } catch (exception: RuntimeException) {
            deleteQuietly(stagedAsset)
            throw exception
        }

        try {
            outbox.create(
                NewOutboxJob(
                    jobId,
                    bot.definition.environment,
                    id,
                    null,
                    encoded.type,
                    null,
                    encoded.payload,
                    now,
                    now,
                    null,
                ),
            )
        } catch (exception: RuntimeException) {
            deleteQuietly(stagedAsset)
            throw exception
        }
        return BotMessageResponse(jobId, false, now)
    }

    private fun deleteQuietly(asset: MediaAsset?) {
        if (asset == null) return
        try {
            mediaStore.delete(asset)
        } catch (_: RuntimeException) {
            // Preserve the request failure; the staged directory remains operator-recoverable.
        }
    }

    private data class EncodedMessage(val type: String, val payload: String)

    private companion object {
        fun requireText(value: String?) {
            if (value.isNullOrBlank()) throw invalid("content is required")
        }

        fun invalid(message: String) =
            BotMessageAdministrationException(HttpStatus.BAD_REQUEST, "INVALID_MESSAGE", message)
    }
}
